SoundWire: mipi-disco: add partial 2.1 support

plbossart opened the pull request:
This PR adds the changes needed to deal with the changes from DisCo 1.0 to 2.1 - without adding any SDCA entity parsing.

The property name changes are handled in a backwards-compatible way.
The most important part is the 'lane-list' property which does impact our multi-lane work, since not all lanes can be used for all ports.

@bardliao please double-check with the MIPI DisCo spec if I missed anything.
@ujfalusi @ranj063 comments welcome

plbossart committed:
ALSA/hda: intel-sdw-acpi: simplify sdw-master-count property read

For some reason we used an array of one u8 when the specification
requires a u32.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ALSA/hda: intel-sdw-acpi: add support for sdw-manager-list property read

The DisCo for SoundWire 2.0 spec adds support for a new
sdw-manager-list property.  Add it in backwards-compatible mode with
'sdw-master-count', which assumed that all links between 0..count-1
exist.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi_disco: add support for clock-scales property

The DisCo for SoundWire 2.0 spec adds support for the
'mipi-sdw-supported-clock-scales' property, which is just a
rename. Add in a backwards-compatible manner.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi-disco: add support for peripheral channelprepare timeout

The DisCo for SoundWire 2.0 spec renamed the
'mipi-sdw-slave-channelprepare-timeout', add support for the new
definition in backwards-compatible ways.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi-disco: add comment on DP0-supported property

The DisCo for SoundWire 2.0 spec adds support for a new property, but
it's not very helpful. Add a comment to explain that it's
intentionally ignored.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi-disco: add new properties from 2.0 spec

The DisCo for SoundWire 2.0 spec adds support for new
'mipi-sdw-sdca-interrupt-register-list' and
'mipi-sdw-commit-register-supported'.

This patch only adds the definitions and property reads, but the use
of these properties will come at some point in the future when needed.

Note a slight conceptual disconnect between the MIPI DisCo definition
of a boolean property and the Linux implementation. The latter only
checks the presence of the property to set its value to 'true',
whereas the MIPI definitions allow for a property with a 'false'
value. This patch relies on a read_u8() even if the DisCo property is
defined as boolean.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi-disco: add support for DP0/DPn 'lane-list' property

The SoundWire specification did not clearly require that ports could
use all Lanes. Some SoundWire/SDCA peripheral adopters added
restrictions on which lanes can be used by what port, and the DisCo
for SoundWire 2.1 specification added a 'lane-list' property to model
this hardware limitation.

When not specified, the ports can use all Lanes. Otherwise, the
'lane-list' indicates which Lanes can be used, sorted by order of
preference (most-preferred-first).

This patch only reads the properties, the use of this property will
come at a later time with multi-lane support.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi-disco: remove DPn audio-modes

The concept of DPn audio-modes was never used by anyone, and was
removed from the DisCo for SoundWire 2.0 specification.

Remove the definitions and TODO.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

ranj063 commented:
looks good to me @plbossart

bardliao commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -79,10 +78,21 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
+			return -EINVAL;
+		}
+
+		list = GENMASK(count - 1, 0);
```
Is it possible that count = 0?

plbossart replied:
Good catch, that would be a problem, I should test this indeed for avoid a nonsensical loop.

ranj063 replied:
but would you even get to the loop, the check for count is right after the if block

plbossart replied:
That's true, we already trap the count = 0. Thanks @ranj063 for the comment.

https://github.com/thesofproject/linux/pull/4857/files/8d44869d3c076940fd95ca751894916b4eae49e6..a3edd2e5dde37524d4f5f103b7c1a6974c7eba02#diff-55d16501c077af0123a7127463a80e233d5cace32c9acf2f26d3289348a9602dR105

So while it's a bit odd to create an all-1 link value, there isn't any bad consequences and the behavior remains the same as before.

bardliao commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -376,6 +376,12 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
+				 &prop->sdca_interrupt_register_list);
+
+	device_property_read_u32(dev, "mipi-sdw-commit-register-supported",
+				 &prop->commit_register_supported);
```
Hmm, `mipi-sdw-commit-register-supported` is a bool.

plbossart replied:
yes, but the implementation of device_property_read_bool() isn't quite right, it only detects the presence of the property and not the value. In the MIPI DisCo spec, it's possible to have a property where 'commit-register-supported' is false.

Adding @andy-shev for more comments on how to deal with this odd configuration.

andy-shev replied:
> yes, but the implementation of device_property_read_bool() isn't quite right, it only detects the presence of the property and not the value.

It's the correct implementation, it's by design like that.  

> In the MIPI DisCo spec, it's possible to have a property where 'commit-register-supported' is false.

The bug is either in the MIPI Disco Specification (if it's really boolean in accordance with the _DSD specifications) **OR** the variable in the code should be an integer (8-, 16-, 32- or 64-bit long).

With that said, the code might need something like this (if the third option is on the table, i.e. property is boolean in terms of MIPI, but _DSD):

    bool foo;
    ...
    u8 value; // but check the exact size
    ...
    value = ...default...;
    device_property_read_uXX(..., &value);
    if (value)
        foo = true;
    else
        foo = false;

plbossart replied:
for my education, where is it specified that a boolean cannot have a 'false' value in the _DSD specs?

I read this in the dsd-guide.pdf page 7 https://github.com/UEFI/DSD-Guide/blob/main/dsd-guide.pdf

````
One use case could be a Boolean value ("yes" means the device has a read-only register that contains a serial number, "no" means the device has no such register)
````

But at the same time, I don't see a Boolean as a valid type in https://uefi.org/sites/default/files/resources/_DSD-device-properties-UUID.pdf

````
The Device Properties UUID
daffd814-6eba-4d8c-8a91-bc9bbf4aa301
defines the data format for the Package (Data Structure) immediately following it as a
list of Packages of length 2 (Properties) where the first element of each contained
Package (Key) must be a String and the second element of it (Value) must be:
• an Integer,
• a String,
• a Reference, or
• a Package consisting entirely of Integer, String, or Reference objects (and
specifically not containing a nested Package).
````

Wondering what the official guidance on Booleans might be?

andy-shev replied:
> for my education, where is it specified that a boolean cannot have a 'false' value in the _DSD specs?
> 
> I read this in the dsd-guide.pdf page 7 https://github.com/UEFI/DSD-Guide/blob/main/dsd-guide.pdf
> 
> ```
> One use case could be a Boolean value ("yes" means the device has a read-only register that contains a serial number, "no" means the device has no such register)
> ```

Oh, this makes a lot of confusion. They are talking about String value where "yes", "no" or other representations are possible to express boolean, but it's not really boolean in terms how it was designed (in the code / mind).

> But at the same time, I don't see a Boolean as a valid type in https://uefi.org/sites/default/files/resources/_DSD-device-properties-UUID.pdf

Of course, there is nothing like that in the ACPI. But we assume (to be compatible with other providers and unification of the API, see below) that boolean is the one which is simply defined by __presence__ of the property, meaning in ACPI it's represented as whatever (in the value) part.  Hence we may test the property:

    device_property_present(...); // gives true or false

and at the same time it means that if we treat the property as boolean, it's presence is enough. Meaning **each** property has (shadowed) boolean equal **true** associated with that.

> Wondering what the official guidance on Booleans might be?

Please, send a patch to that repository, it's indeed unclear at all. And it's definitely a documentation bug.

You also may come to OS Firmware meeting (next will be tomorrow) and talk about this issue. Might make sense to summon Sakari Ailus to the discussion.

plbossart replied:
@andy-shev the DSDT of the device I have on my desk contains this:

````
Package (0x02)                                                            
{                                                                         
    "mipi-sdw-commit-register-supported",                                 
    Zero                                                                  
},
````

So at this point the cat is out of the bag.

andy-shev replied:
@plbossart It has to be handled as integer (let's say u8, but you need to see in the specification for the range of values) **if value matters**.

plbossart replied:
used u8.

vijendarmukunda commented:
LGTM

vijendarmukunda reviewed (commented):
I think sdw_intel_scan_controller() function can be made as generic . Only delta we see max no of links. If we handle it correctly, then I think we can use the same function sdw_intel_scan_controller() instead of  referring amd_sdw_scan_controller() for AMD platforms.

vijendarmukunda commented:
> I think sdw_intel_scan_controller() function can be made as generic . Only delta we see max no of links. If we handle it correctly, then I think we can use the same function sdw_intel_scan_controller() instead of referring amd_sdw_scan_controller() for AMD platforms.

May be we can go with generic function keeping max links value same for Intel and AMD platforms. We can take care invalid channel mask value in amd_init.c file.

plbossart commented:
> > I think sdw_intel_scan_controller() function can be made as generic . Only delta we see max no of links. If we handle it correctly, then I think we can use the same function sdw_intel_scan_controller() instead of referring amd_sdw_scan_controller() for AMD platforms.
> 
> May be we can go with generic function keeping max links value same for Intel and AMD platforms. We can take care invalid channel mask value in amd_init.c file.

Not sure how to go about sharing, AMD and Intel have both vendor-specific properties. We'd need some sort of callback or have a routine where we read all properties

vijendarmukunda commented:
> > > I think sdw_intel_scan_controller() function can be made as generic . Only delta we see max no of links. If we handle it correctly, then I think we can use the same function sdw_intel_scan_controller() instead of referring amd_sdw_scan_controller() for AMD platforms.
> > 
> > 
> > May be we can go with generic function keeping max links value same for Intel and AMD platforms. We can take care invalid channel mask value in amd_init.c file.
> 
> Not sure how to go about sharing, AMD and Intel have both vendor-specific properties. We'd need some sort of callback or have a routine where we read all properties

Agreed.

bardliao commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -284,6 +284,34 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+			continue;
+		len = strlen(prop_val);
+		/* The last character is enough to identify the connection */
+		prop->lane_maps[i] = prop_val[len - 1];
```
Should this be `prop->lane_maps[i] = prop_val[len - 1] - '0';`? `prop_val[len - 1]` is a char and the value is '1' to '7'. And we treat prop->lane_maps[] as integer, right?

plbossart replied:
I don't know, I took Chao's code as is...

andy-shev replied:
> Should this be `prop->lane_maps[i] = prop_val[len - 1] - '0';`? `prop_val[len - 1]` is a char and the value is '1' to '7'. And we treat prop->lane_maps[] as integer, right?

Do we have `isodigit()`? Or `isdigit()` can be used.

bardliao replied:
> I don't know, I took Chao's code as is...

I would suggest that we treat prop->lane_maps[] as integer, and convert '1' to 1 here.

bardliao replied:
> > Should this be `prop->lane_maps[i] = prop_val[len - 1] - '0';`? `prop_val[len - 1]` is a char and the value is '1' to '7'. And we treat prop->lane_maps[] as integer, right?
> 
> Do we have `isodigit()`? Or `isdigit()` can be used.

@andy-shev Are you suggesting we check `isdigit(prop_val[len - 1])` before ` prop->lane_maps[i] = prop_val[len - 1] - '0';`?

andy-shev replied:
> > > Should this be `prop->lane_maps[i] = prop_val[len - 1] - '0';`? `prop_val[len - 1]` is a char and the value is '1' to '7'. And we treat prop->lane_maps[] as integer, right?
> > 
> > Do we have `isodigit()`? Or `isdigit()` can be used.
> 
> @andy-shev Are you suggesting we check `isdigit(prop_val[len - 1])` before ` prop->lane_maps[i] = prop_val[len - 1] - '0';`?

There are many ways of converting that. Once can use `is*()` API to check, or directly call something like `hex_to_bin()` first (from hex.h) and compare result to be in a range: `in_range(hex_to_bin(...), ...)`.
You may also leave the above, but it might be less illustrative on what's going on there.

If you want more robust code, you probably want `ret = kstrtou8(...);  if (ret) {...handle error...} if (in_range(...handle range...) {}`. This will allow to support 2 figure numbers if needed (in the future?).

plbossart replied:
patch dropped, will the added by the multi-lane work in PR #4733

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -284,6 +284,34 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+		ret = device_property_read_string(dev, prop_name, &prop_val);
+		if (ret)
+			continue;
+		len = strlen(prop_val);
```
IIRC, the string can be empty, this code then has a potential out-of-boundary access.

plbossart commented on include/linux/soundwire/sdw.h:
```diff
@@ -283,6 +287,8 @@ struct sdw_dp0_prop {
  * @read_only_wordlength: Read Only wordlength field in DPN_BlockCtrl1 register
  * @port_encoding: Payload Channel Sample encoding schemes supported
+ * @num_lanes: array size of @lane_list
+ * @lane_list: indicates which Lanes can be used by DPn
```
@bardliao note that this new property will constrain your multi-lane work. It's not enough to check if all peripherals are connected on a shared lane, we need **in addition** to make sure there's an available port to read/write from that lane.

bardliao commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -57,16 +57,17 @@ static int
-	u8 count, i;
+	u32 count;
+	int i;
 	int ret;
```
nitpick: Would it be better to move `int i;` after `int ret;`?
Like:
```
u32 count;
int ret;
int i;
```

andy-shev replied:
And I even go further to have `unsigned int i;` in the code.

plbossart replied:
done

bardliao commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -57,19 +57,18 @@ static int
 {
 	struct acpi_device *adev = acpi_fetch_acpi_dev(info->handle);
+	unsigned long list_ul;
+	u32 list;
```
nitpick: And here, `u32 list;` after `u32 count;`

plbossart replied:
done

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -57,16 +57,17 @@ static int
 	count = 0;
-	ret = fwnode_property_read_u8_array(acpi_fwnode_handle(adev),
-					    "mipi-sdw-master-count", &count, 1);
+	ret = fwnode_property_read_u32(acpi_fwnode_handle(adev),
```
I'm not sure how `ret` is being used, but I assume the above `count=0;` can be done conditionally. OR the `ret = ` part completely dropped (in case this is optional).

plbossart replied:
@andy-shev the 'ret' is used below line 84.

I am not sure if it's good practice to pass an un-initialized value to the fwnode_property_read_u32. I don't know what the fwnode code does...

This pattern has been around for about 5 years now, so it's not really new by any stretch of imagination.

andy-shev replied:
If you check the returned value, you may already know that the result is undefined and assign it accordingly. So, yes, it's good to pass just a pointer to uninitialised value for the mandatory properties (means the code checks for errors returned by fwnode APIs).

plbossart replied:
done

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -81,10 +80,21 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
-		dev_err(&adev->dev,
-			"Failed to read mipi-sdw-master-count: %d\n", ret);
-		return -EINVAL;
+	count = 0;
```
Same Q, how is this being used?

plbossart replied:
prudent initialization since I don't know what the fwnode code does. It's not new code...

andy-shev replied:
Seems unneeded churn, but I don't know the code, you should decide.

plbossart replied:
removed

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -81,10 +80,21 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
+		if (ret) {
+			dev_err(&adev->dev,
+				"Failed to read mipi-sdw-manager-list or mipi-sdw-master-count: %d\n", ret);
+			return -EINVAL;
```
Please, unshadow the error code.

plbossart replied:
that would be a separate change, see on the left that's what the code did before.

plbossart replied:
done

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -81,10 +80,21 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
+	count = 0;
+	ret = fwnode_property_read_u32(acpi_fwnode_handle(adev),
+				       "mipi-sdw-manager-list", &list);
+	if (ret < 0) {
```
Why ` < 0` ?

plbossart replied:
not sure why, it's indeed inconsistent.

plbossart replied:
used if (ret)

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -81,10 +80,21 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
+	ret = fwnode_property_read_u32(acpi_fwnode_handle(adev),
+				       "mipi-sdw-manager-list", &list);
+	if (ret < 0) {
+		ret = fwnode_property_read_u32(acpi_fwnode_handle(adev),
```
This can be one level of indentation less.

plbossart replied:
done

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -102,8 +112,9 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
 
 	info->count = count;
 	info->link_mask = 0;
+	list_ul = list;
```
Can't list be converted to `unsigned long`?

plbossart replied:
see above, we're using  hweight32(list);

andy-shev replied:
I don't buy this, but I see that it's retrieved as `u32` from the properties. Nevertheless, I just commented there that would be better to have a helper that returns a `list` bitmask which is `unsigned long` by default.

plbossart replied:
used the try_count/skip_count suggestion, thanks.

Left the u32 list as is.

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -102,8 +112,9 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
 
-	for (i = 0; i < count; i++) {
+	for_each_set_bit(i, &list_ul, SDW_MAX_LINKS) {
 		if (ctrl_link_mask && !(ctrl_link_mask & BIT(i))) {
```
For the future: this can be checked before this loop via bitmap operations, like
```
bitmap_andnot(tmp, list, _link_mask, _MAX_LINKS);
if (bitmap_intersects(tmp, _link_mask, _MAX_LINKS)) {
  bitmap_andnot(list, list, tmp, _MAX_LINKS);
  dev_dbg("%pb\n", tmp);
}
for_each_set_bit(i, list, SDW_MAX_LINKS) {
...
}
```
if I am not mistaken.

plbossart replied:
There are two conditions inside the loop
a) the bitmask is set
b) the link is enabled/disabled

Best to keep the two in the same scope for readability, no?

andy-shev replied:
Yeah, maybe better your way. I gave a proposal for the consideration.

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -100,7 +106,7 @@ int sdw_master_read_prop(struct sdw_bus *bus)
 
 		fwnode_property_read_u32_array(link,
-					       "mipi-sdw-supported-clock-gears",
+					       scales_prop,
```
Can be moved to the previous line now.

plbossart replied:
Yes, but then so can the two others below. I chose to minimize changes for readability of the diff, but arguably the code could be improved with fewer lines used.

andy-shev replied:
Okay.

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -88,7 +89,12 @@ int sdw_master_read_prop(struct sdw_bus *bus)
 	}
 
-	nval = fwnode_property_count_u32(link, "mipi-sdw-supported-clock-gears");
+	scales_prop = "mipi-sdw-supported-clock-scales";
```
Hmm... Personally I would extract a helper which will be called instead with the property name as a parameter (or even the differential part of it: `"gears"` or `"scales"`).

plbossart replied:
not following, the helper IS fwnode_property_count_u32().
We need to try the first property name and if not found try the other one.

andy-shev replied:
I meant something on top of that. But on the second though it would become an unreadable as we partially hide the property name. So, scratch that idea.

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -330,8 +331,11 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
 
-	device_property_read_u32(dev, "mipi-sdw-slave-channelprepare-timeout",
-				 &prop->ch_prep_timeout);
+	ret = device_property_read_u32(dev, "mipi-sdw-peripheral-channelprepare-timeout",
```
Yeah, not the first time and most likely not the last we will see similar fallbacks. Perhaps you can extend the generic device property API to support this? Like you supply the list of the possible properties in the order of high to low priority
```
static inline int device_property_read_u32_one_of(..., const char * const *properties, u32 *value)
{
  const char *p;
  int ret;

  for (p = *properties; p; p++) {
    ret = ...;
    if (ret != ...) // code for no property found
      return ret;
  }

  return ...; // same code as per above
}
```
either as inliners or macros (with a type passed through `_Generic()`)?

plbossart replied:
Humm, not sure. This would work when the change is only a renaming, but the parameters and meaning mean the same. But when the value changes between a count and a bitmask, you need to know which of the two properties was found to deal with error handling.

andy-shev replied:
I see what you mean. I even have a real case as you described, but still we may return an index of the property that were found and parsed: < 0 — an error, >= 0 — an index in property names array, from which one may decide how to interpret the result.

plbossart replied:
I chose to keep the code as is. This case was an exception where the DisCo spec changed the name AND the arguments, that was an editing mistake that should have been caught during the spec review.

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -362,7 +362,11 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
+	/*
+	 * Read dp0 properties - we don't rely on the 'mipi-sdw-dp-0-supported'
+	 * property since the 'mipi-sdw-dp0-subproperties' property is logically
+	 * equivalent
```
Missing period at the end.

plbossart replied:
done

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -268,7 +268,7 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
 					GFP_KERNEL);
 			if (!dpn[i].ch_combinations) {
-				fwnode_handle_put(node);						
+				fwnode_handle_put(node);
```
What has happened here?

plbossart replied:
probably I removed a trailing whitespace that was there before...

plbossart replied:
indeed, I need to rebase all this after correcting the initial mistake.

plbossart replied:
fixed

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -362,6 +362,19 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
+	 * the value can be defined as zero. This is not aligned the
+	 * implementation of device_property_read_bool() which only checks
+	 * the presence of the property.
+	 * Let's use read_u8 to work-around this conceptual disconnect
```
Missing period.

plbossart replied:
done

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -168,6 +168,20 @@ static int sdw_slave_read_dp0(struct sdw_slave *slave,
+		if (!dp0->lane_list)
+			return -ENOMEM;
+
+		fwnode_property_read_u32_array(port,
```
No error check? What can go wrong?.. :-)

plbossart replied:
well, this is a specific case @andy-shev. We first read the array to figure out its element count, then read a second time to copy the contents in the allocated array. The existing code for SoundWire assumes the second read completes without any errors, and for consistency I used the same programming pattern.

If this is wrong then we need to correct the entire file...

andy-shev replied:
This __might__ be wrong. Basically there is no guarantee that the second read will succeed. For example, `asprinf()`-based functions in kernel have a specific warning in case the string (length) had been changed between two calls. It means, there is no guarantee that similar never happen here. TL;DR: I suggest considering such a patch.

plbossart replied:
yeah, but here we're reading fixed information set in stone in the DSDT... The DisCo stuff is so horrible no one in their right mind will try to do something dynamic. The DisCo tool generate a file that's not supposed to be modified, ever.

andy-shev replied:
I understand that, but what I'm talking about is the OS part (data corruption, other thingy), which DisCo tool and ACPI has no control over. Additionally, but I believe it's pure theoretical, some of that might be instantiated via SSDT overlay, which **is** dynamic ACPI piece of namespace.

plbossart replied:
error handling added.

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -290,6 +304,20 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+			if (!dpn[i].lane_list)
+				return -ENOMEM;
+
+			fwnode_property_read_u32_array(node,
```
As per above.

plbossart replied:
error handling added

plbossart committed:
ALSA/hda: intel-sdw-acpi: cleanup sdw_intel_scan_controller

Remove unnecessary initialization and un-shadow return code.

Suggested-by: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: mipi-disco: add error handling for property array read

The existing code assumes that there are no possible errors when using
fwnode_property_read_u32_array(), because fwnode_property_count_u32()
reads this array to determine its number of elements. We need to also
protect the second read to be completely bullet-proof.

Suggested-by: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -378,6 +378,19 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
+				 &prop->sdca_interrupt_register_list);
+
+	/*
+	 * the specification defines the property value as boolean, but
```
"The specification..."

plbossart replied:
done, thanks.

andy-shev commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -266,10 +280,12 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+				return ret;
 		}
 
 		fwnode_property_read_u32(node,
```
Here as well? But looking at the code it seems a lot of such calls are done without checks, so let it be for now (i.e. only `_array()` ones to check) as it most likely requires a good audit.

plbossart replied:
Yeah, I thought about hardening the error-handling, but the thing is that most DSDTs don't include all the properties, so if we bail then we might prevent the probe from happening. That's really not desired.

In a perfect world, OEMs would have included all the properties. In reality, even the Intel reference is incomplete.

andy-shev replied:
As I said, let it be then.

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -79,13 +79,25 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
 	 *
 	 * We will check the hardware capabilities in the startup() step
 	 */
+	ret = fwnode_property_read_u32(acpi_fwnode_handle(adev),
```
Perhaps

```
struct fwnode_handle *fwnode = acpi_fwnode_handle(adev);
```
and reuse?

plbossart replied:
done

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -79,13 +79,25 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
 	 */
+	ret = fwnode_property_read_u32(acpi_fwnode_handle(adev),
+				       "mipi-sdw-manager-list", &list);
+	if (!ret) {
```
I would think more of the usual pattern
```
  if (ret)
    goto try_count;

  count = ...;
  goto skip_count;
```
Or having this as a helper which will return it for you
```
ret = _get_list_of(..., &list);
if (ret)
 ...
```
In this case you can get rid of `list_ul`.

plbossart committed:
ALSA/hda: intel-sdw-acpi: fetch fwnode once in sdw_intel_scan_controller()

Optimize a bit by using an intermediate 'fwnode' variable.

Suggested-by: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

andy-shev commented on sound/hda/intel-sdw-acpi.c:
```diff
@@ -58,19 +58,20 @@ sdw_intel_scan_controller(struct sdw_intel_acpi_info *info)
 {
 	struct acpi_device *adev = acpi_fetch_acpi_dev(info->handle);
 	struct fwnode_handle *fwnode;
+	unsigned long list_ul;
```
I believe the same comment applies here, i.e. splitting out the helper that will return `list` in `unsigned long` form. This will make code cleaner. But it might be that you do in the sequential patches (haven't seen them yet).

plbossart replied:
coming back to this after a month debugging other things, I don't really remember or understand what you meant @andy-shev ?

andy-shev replied:
I meant that instead of adding a new variable you can make a helper that will return an unsigned long for all cases and use bit ops on top of that. When I come to work after vacations (mid-August) I might be able to draft it.

plbossart replied:
ok, I won't have the time to work on this so will wait for your suggestion @andy-shev.

andy-shev replied:
> ok, I won't have the time to work on this so will wait for your suggestion @andy-shev.

Sure, just ping me then.

charleskeepax replied:
Personally I think the problem really is the nest of gotos, not sure a helper function is strictly necessary, although fine by me. I would probably suggest just drop the inclusion of both strings in the error message, which I assume is driving the gotos. I would probably also move the conversion to unsigned long up to the parsing as well something like:

        unsigned long list;
        u32 count, tmp;

        ret = fwnode_property_read_u32(fwnode, "mipi-sdw-manager-list", &tmp);
        if (ret) {
                ret = fwnode_property_read_u32(fwnode, "mipi-sdw-master-count", &count);
                if (ret) {
                        dev_err(&adev->dev, "Failed to read mipi-sdw-manager-list: %d\n", ret);
                        return ret;
                }

                list = GENMASK(count - 1, 0);
        } else {
                list = tmp;
                count = hweight32(list);
        }

Rest of this series looks good to me though.

andy-shev replied:
> Personally I think the problem really is the nest of gotos, not sure a helper function is strictly necessary, although fine by me. I would probably suggest just drop the inclusion of both strings in the error message, which I assume is driving the gotos. I would probably also move the conversion to unsigned long up to the parsing as well something like: ...

LGTM.

plbossart replied:
changed the code as suggested, thanks for the feedback @andy-shev @charleskeepax

andy-shev commented on include/linux/soundwire/sdw.h:
```diff
@@ -366,6 +369,8 @@ struct sdw_slave_prop {
 	u32 quirks;
 	bool clock_reg_supported;
 	bool use_domain_irq;
+	u32 sdca_interrupt_register_list;
```
I would recommend to run `pahole` and perhaps rearrange the fields to remove unnecessary gaps and reduce memory footprint.

plbossart replied:
done

andy-shev commented on include/linux/soundwire/sdw.h:
```diff
@@ -252,6 +254,8 @@ struct sdw_dp0_prop {
 	bool simple_ch_prep_sm;
 	u32 ch_prep_timeout;
 	bool imp_def_interrupts;
+	int num_lanes;
```
Same, basically the rule of thumb is to run `pahole` each time after updating the structure (data type) and see if it can be improved.

plbossart replied:
I don't think anyone has ever used this tool on the SOF driver. I just tried for 15mn without any success, and it's just yet another possibly great tool that I'll never remember come next week. Oh well.

andy-shev replied:
> I don't think anyone has ever used this tool on the SOF driver. I just tried for 15mn without any success, and it's just yet another possibly great tool that I'll never remember come next week. Oh well.

Should be easy to use...

plbossart replied:
we didn't have the right Kconfig options, works fine now

https://github.com/thesofproject/kconfig/pull/98

plbossart replied:
done

andy-shev commented on include/linux/soundwire/sdw.h:
```diff
@@ -306,6 +312,8 @@ struct sdw_dpn_prop {
 	bool block_pack_mode;
 	bool read_only_wordlength;
 	u32 port_encoding;
+	int num_lanes;
```
Ditto.

plbossart replied:
done

plbossart committed:
soundwire: optimize sdw_stream_runtime memory layout

pahole suggestion: swap position of 'm_rt_count'

before: pahole -C sdw_stream_runtime drivers/soundwire/soundwire-bus.ko
struct sdw_stream_runtime {
	const char  *              name;                 /*     0     8 */
	struct sdw_stream_params   params;               /*     8    12 */
	enum sdw_stream_state      state;                /*    20     4 */
	enum sdw_stream_type       type;                 /*    24     4 */

	/* XXX 4 bytes hole, try to pack */

	struct list_head           master_list;          /*    32    16 */
	int                        m_rt_count;           /*    48     4 */

	/* size: 56, cachelines: 1, members: 6 */
	/* sum members: 48, holes: 1, sum holes: 4 */
	/* padding: 4 */
	/* last cacheline: 56 bytes */
};

after: pahole --reorganize -C sdw_stream_runtime drivers/soundwire/soundwire-bus.ko
struct sdw_stream_runtime {
	const char  *              name;                 /*     0     8 */
	struct sdw_stream_params   params;               /*     8    12 */
	enum sdw_stream_state      state;                /*    20     4 */
	enum sdw_stream_type       type;                 /*    24     4 */
	int                        m_rt_count;           /*    28     4 */
	struct list_head           master_list;          /*    32    16 */

	/* size: 48, cachelines: 1, members: 6 */
	/* last cacheline: 48 bytes */
};   /* saved 8 bytes! */

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: optimize sdw_bus structure

The sdw_bus structure has seen multiple additions over the years. It's
one of the most used structures in this subsystem, so there's merit in
reshuffling the members a bit with 'pahole' to reduce holes and
structures across cache lines.

before:

struct sdw_bus {
	struct device *            dev;                  /*     0     8 */
	struct sdw_master_device * md;                   /*     8     8 */
	int                        controller_id;        /*    16     4 */
	unsigned int               link_id;              /*    20     4 */
	int                        id;                   /*    24     4 */

	/* XXX 4 bytes hole, try to pack */

	struct list_head           slaves;               /*    32    16 */
	long unsigned int          assigned[1];          /*    48     8 */
	struct mutex               bus_lock;             /*    56   160 */
	/* --- cacheline 3 boundary (192 bytes) was 24 bytes ago --- */
	struct lock_class_key      bus_lock_key;         /*   216    16 */
	struct mutex               msg_lock;             /*   232   160 */
	/* --- cacheline 6 boundary (384 bytes) was 8 bytes ago --- */
	struct lock_class_key      msg_lock_key;         /*   392    16 */
	int                        (*compute_params)(struct sdw_bus *); /*   408     8 */
	const struct sdw_master_ops  * ops;              /*   416     8 */
	const struct sdw_master_port_ops  * port_ops;    /*   424     8 */
	struct sdw_bus_params      params;               /*   432    36 */

	/* XXX 4 bytes hole, try to pack */

	/* --- cacheline 7 boundary (448 bytes) was 24 bytes ago --- */
	struct sdw_master_prop     prop;                 /*   472    72 */

	/* XXX last struct has 6 bytes of padding */

	/* --- cacheline 8 boundary (512 bytes) was 32 bytes ago --- */
	void *                     vendor_specific_prop; /*   544     8 */
	struct list_head           m_rt_list;            /*   552    16 */
	struct dentry *            debugfs;              /*   568     8 */
	/* --- cacheline 9 boundary (576 bytes) --- */
	struct irq_chip            irq_chip;             /*   576   264 */
	/* --- cacheline 13 boundary (832 bytes) was 8 bytes ago --- */
	struct irq_domain *        domain;               /*   840     8 */
	struct sdw_defer           defer_msg;            /*   848   112 */
	/* --- cacheline 15 boundary (960 bytes) --- */
	unsigned int               clk_stop_timeout;     /*   960     4 */
	u32                        bank_switch_timeout;  /*   964     4 */
	bool                       multi_link;           /*   968     1 */

	/* XXX 3 bytes hole, try to pack */

	int                        hw_sync_min_links;    /*   972     4 */
	int                        stream_refcount;      /*   976     4 */

	/* size: 984, cachelines: 16, members: 27 */
	/* sum members: 969, holes: 3, sum holes: 11 */
	/* padding: 4 */
	/* paddings: 1, sum paddings: 6 */
	/* last cacheline: 24 bytes */
};

after:

struct sdw_bus {
	struct device *            dev;                  /*     0     8 */
	struct sdw_master_device * md;                   /*     8     8 */
	struct lock_class_key      bus_lock_key;         /*    16    16 */
	struct mutex               bus_lock;             /*    32   160 */
	/* --- cacheline 3 boundary (192 bytes) --- */
	struct list_head           slaves;               /*   192    16 */
	struct lock_class_key      msg_lock_key;         /*   208    16 */
	struct mutex               msg_lock;             /*   224   160 */
	/* --- cacheline 6 boundary (384 bytes) --- */
	struct list_head           m_rt_list;            /*   384    16 */
	struct sdw_defer           defer_msg;            /*   400   112 */
	/* --- cacheline 8 boundary (512 bytes) --- */
	struct sdw_bus_params      params;               /*   512    36 */
	int                        stream_refcount;      /*   548     4 */
	const struct sdw_master_ops  * ops;              /*   552     8 */
	const struct sdw_master_port_ops  * port_ops;    /*   560     8 */
	struct sdw_master_prop     prop;                 /*   568    72 */

	/* XXX last struct has 6 bytes of padding */

	/* --- cacheline 10 boundary (640 bytes) --- */
	void *                     vendor_specific_prop; /*   640     8 */
	int                        hw_sync_min_links;    /*   648     4 */
	int                        controller_id;        /*   652     4 */
	unsigned int               link_id;              /*   656     4 */
	int                        id;                   /*   660     4 */
	int                        (*compute_params)(struct sdw_bus *); /*   664     8 */
	long unsigned int          assigned[1];          /*   672     8 */
	unsigned int               clk_stop_timeout;     /*   680     4 */
	u32                        bank_switch_timeout;  /*   684     4 */
	struct irq_chip            irq_chip;             /*   688   264 */
	/* --- cacheline 14 boundary (896 bytes) was 56 bytes ago --- */
	struct irq_domain *        domain;               /*   952     8 */
	/* --- cacheline 15 boundary (960 bytes) --- */
	struct dentry *            debugfs;              /*   960     8 */
	bool                       multi_link;           /*   968     1 */

	/* size: 976, cachelines: 16, members: 27 */
	/* padding: 7 */
	/* paddings: 1, sum paddings: 6 */
	/* last cacheline: 16 bytes */
};

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: optimize sdw_master_prop

Make pahole happy by moving pointers and u64 first instead of
interleaving them.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: optimize sdw_slave_prop

move pointers first, and move booleans together.

before:

struct sdw_slave_prop {
	u32                        mipi_revision;        /*     0     4 */
	bool                       wake_capable;         /*     4     1 */
	bool                       test_mode_capable;    /*     5     1 */
	bool                       clk_stop_mode1;       /*     6     1 */
	bool                       simple_clk_stop_capable; /*     7     1 */
	u32                        clk_stop_timeout;     /*     8     4 */
	u32                        ch_prep_timeout;      /*    12     4 */
	enum sdw_clk_stop_reset_behave reset_behave;     /*    16     4 */
	bool                       high_PHY_capable;     /*    20     1 */
	bool                       paging_support;       /*    21     1 */
	bool                       bank_delay_support;   /*    22     1 */

	/* XXX 1 byte hole, try to pack */

	enum sdw_p15_behave        p15_behave;           /*    24     4 */
	bool                       lane_control_support; /*    28     1 */

	/* XXX 3 bytes hole, try to pack */

	u32                        master_count;         /*    32     4 */
	u32                        source_ports;         /*    36     4 */
	u32                        sink_ports;           /*    40     4 */

	/* XXX 4 bytes hole, try to pack */

	struct sdw_dp0_prop *      dp0_prop;             /*    48     8 */
	struct sdw_dpn_prop *      src_dpn_prop;         /*    56     8 */
	/* --- cacheline 1 boundary (64 bytes) --- */
	struct sdw_dpn_prop *      sink_dpn_prop;        /*    64     8 */
	u8                         scp_int1_mask;        /*    72     1 */

	/* XXX 3 bytes hole, try to pack */

	u32                        quirks;               /*    76     4 */
	bool                       clock_reg_supported;  /*    80     1 */
	bool                       use_domain_irq;       /*    81     1 */

	/* size: 88, cachelines: 2, members: 23 */
	/* sum members: 71, holes: 4, sum holes: 11 */
	/* padding: 6 */
	/* last cacheline: 24 bytes */
};

after:

truct sdw_slave_prop {
	struct sdw_dp0_prop *      dp0_prop;             /*     0     8 */
	struct sdw_dpn_prop *      src_dpn_prop;         /*     8     8 */
	struct sdw_dpn_prop *      sink_dpn_prop;        /*    16     8 */
	u32                        mipi_revision;        /*    24     4 */
	bool                       wake_capable;         /*    28     1 */
	bool                       test_mode_capable;    /*    29     1 */
	bool                       clk_stop_mode1;       /*    30     1 */
	bool                       simple_clk_stop_capable; /*    31     1 */
	u32                        clk_stop_timeout;     /*    32     4 */
	u32                        ch_prep_timeout;      /*    36     4 */
	enum sdw_clk_stop_reset_behave reset_behave;     /*    40     4 */
	bool                       high_PHY_capable;     /*    44     1 */
	bool                       paging_support;       /*    45     1 */
	bool                       bank_delay_support;   /*    46     1 */
	bool                       lane_control_support; /*    47     1 */
	enum sdw_p15_behave        p15_behave;           /*    48     4 */
	u32                        master_count;         /*    52     4 */
	u32                        source_ports;         /*    56     4 */
	u32                        sink_ports;           /*    60     4 */
	/* --- cacheline 1 boundary (64 bytes) --- */
	u32                        quirks;               /*    64     4 */
	u8                         scp_int1_mask;        /*    68     1 */
	bool                       clock_reg_supported;  /*    69     1 */
	bool                       use_domain_irq;       /*    70     1 */

	/* size: 72, cachelines: 2, members: 23 */
	/* padding: 1 */
	/* last cacheline: 8 bytes */
};

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: optimize sdw_dp0_prop

Move pointers and booleans.

Before:

struct sdw_dp0_prop {
	u32                        max_word;             /*     0     4 */
	u32                        min_word;             /*     4     4 */
	u32                        num_words;            /*     8     4 */

	/* XXX 4 bytes hole, try to pack */

	u32 *                      words;                /*    16     8 */
	bool                       BRA_flow_controlled;  /*    24     1 */
	bool                       simple_ch_prep_sm;    /*    25     1 */

	/* XXX 2 bytes hole, try to pack */

	u32                        ch_prep_timeout;      /*    28     4 */
	bool                       imp_def_interrupts;   /*    32     1 */

	/* size: 40, cachelines: 1, members: 8 */
	/* sum members: 27, holes: 2, sum holes: 6 */
	/* padding: 7 */
	/* last cacheline: 40 bytes */
};

after:

struct sdw_dp0_prop {
	u32 *                      words;                /*     0     8 */
	u32                        max_word;             /*     8     4 */
	u32                        min_word;             /*    12     4 */
	u32                        num_words;            /*    16     4 */
	u32                        ch_prep_timeout;      /*    20     4 */
	bool                       BRA_flow_controlled;  /*    24     1 */
	bool                       simple_ch_prep_sm;    /*    25     1 */
	bool                       imp_def_interrupts;   /*    26     1 */

	/* size: 32, cachelines: 1, members: 8 */
	/* padding: 5 */
	/* last cacheline: 32 bytes */
};

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: optimize sdw_dpn_prop

before:
struct sdw_dpn_prop {
	u32                        num;                  /*     0     4 */
	u32                        max_word;             /*     4     4 */
	u32                        min_word;             /*     8     4 */
	u32                        num_words;            /*    12     4 */
	u32 *                      words;                /*    16     8 */
	enum sdw_dpn_type          type;                 /*    24     4 */
	u32                        max_grouping;         /*    28     4 */
	bool                       simple_ch_prep_sm;    /*    32     1 */

	/* XXX 3 bytes hole, try to pack */

	u32                        ch_prep_timeout;      /*    36     4 */
	u32                        imp_def_interrupts;   /*    40     4 */
	u32                        max_ch;               /*    44     4 */
	u32                        min_ch;               /*    48     4 */
	u32                        num_channels;         /*    52     4 */
	u32 *                      channels;             /*    56     8 */
	/* --- cacheline 1 boundary (64 bytes) --- */
	u32                        num_ch_combinations;  /*    64     4 */

	/* XXX 4 bytes hole, try to pack */

	u32 *                      ch_combinations;      /*    72     8 */
	u32                        modes;                /*    80     4 */
	u32                        max_async_buffer;     /*    84     4 */
	bool                       block_pack_mode;      /*    88     1 */
	bool                       read_only_wordlength; /*    89     1 */

	/* XXX 2 bytes hole, try to pack */

	u32                        port_encoding;        /*    92     4 */
	struct sdw_dpn_audio_mode * audio_modes;         /*    96     8 */

	/* size: 104, cachelines: 2, members: 22 */
	/* sum members: 95, holes: 3, sum holes: 9 */
	/* last cacheline: 40 bytes */
};

after:

struct sdw_dpn_prop {
	struct sdw_dpn_audio_mode * audio_modes;         /*     0     8 */
	u32                        num;                  /*     8     4 */
	u32                        max_word;             /*    12     4 */
	u32                        min_word;             /*    16     4 */
	u32                        num_words;            /*    20     4 */
	u32 *                      words;                /*    24     8 */
	enum sdw_dpn_type          type;                 /*    32     4 */
	u32                        max_grouping;         /*    36     4 */
	u32                        ch_prep_timeout;      /*    40     4 */
	u32                        imp_def_interrupts;   /*    44     4 */
	u32                        max_ch;               /*    48     4 */
	u32                        min_ch;               /*    52     4 */
	u32                        num_channels;         /*    56     4 */
	u32                        num_ch_combinations;  /*    60     4 */
	/* --- cacheline 1 boundary (64 bytes) --- */
	u32 *                      channels;             /*    64     8 */
	u32 *                      ch_combinations;      /*    72     8 */
	u32                        modes;                /*    80     4 */
	u32                        max_async_buffer;     /*    84     4 */
	u32                        port_encoding;        /*    88     4 */
	bool                       block_pack_mode;      /*    92     1 */
	bool                       read_only_wordlength; /*    93     1 */
	bool                       simple_ch_prep_sm;    /*    94     1 */

	/* size: 96, cachelines: 2, members: 22 */
	/* padding: 1 */
	/* last cacheline: 32 bytes */
};

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
new version: rebased, patches reordered and minor change suggested by @charleskeepax

bardliao commented:
All comments are addressed. Let's merge it.
